script.kodi.windows.update/default.py

```diff
@@ -0,0 +1,249 @@
+            url = newURL.replace('Master','master') + '/'
+            
+            
+    def fileExists(self, dest):
+        if xbmcvfs.exists(dest):
+            if not xbmcgui.Dialog().yesno(ADDON_NAME, LANGUAGE(30004), dest.rsplit('/', 1)[-1], nolabel=LANGUAGE(30005), yeslabel=LANGUAGE(30006)): return False
+        elif CLEAN and xbmcvfs.exists(self.lastPath): self.deleteEXE(self.lastPath)
+        return False
+        
+        
+    def deleteEXE(self, path):
+        #some file systems don't release the file lock instantly.
+        for count in range(3):
+            if self.myMonitor.waitForAbort(1): return 
+            try: 
+                if xbmcvfs.delete(path): return
+            except: pass
+        
+        
+    def downloadEXE(self, url, dest):
+        if self.fileExists(dest): return self.installEXE(dest)
+        start_time = time.time()
+        dia = xbmcgui.DialogProgress()
+        fle = dest.rsplit('\\', 1)[1]
+        dia.create(ADDON_NAME, LANGUAGE(30002))
+        try: urllib.request.urlretrieve(url.rstrip('/'), dest, lambda nb, bs, fs: self.pbhook(nb, bs, fs, dia, start_time, fle))
+        except Exception as e:
+            dia.close()
+            xbmcgui.Dialog().notification(ADDON_NAME, LANGUAGE(30001), ICON, 4000)
+            log("downloadAPK, Failed! (%s) %s"%(url,str(e)), xbmc.LOGERROR)
+            return self.deleteEXE(dest)
+        return self.installEXE(dest)
+
+
+    def pbhook(self, numblocks, blocksize, filesize, dia, start_time, fle):
+        try: 
+            percent = min(numblocks * blocksize * 100 / filesize, 100) 
+            currently_downloaded = float(numblocks) * blocksize / (1024 * 1024) 
+            kbps_speed = numblocks * blocksize / (time.time() - start_time) 
+            if kbps_speed > 0: eta = (filesize - numblocks * blocksize) / kbps_speed 
+            else: eta = 0 
+            kbps_speed = kbps_speed / 1024 
+            if eta < 0: eta = divmod(0, 60)
+            else: eta = divmod(eta, 60)
+            total   = (float(filesize) / (1024 * 1024))
+            label   = '[B]Downloading:[/B] %s'%SETTINGS_LOC
+            label2  = '%.02f MB of %.02f MB'%(currently_downloaded,total)
+            label2 += ' | [B]Speed:[/B] %.02f Kb/s'%kbps_speed
+            label2 += ' | [B]ETA:[/B] %02d:%02d'%eta
+            dia.update(int(percent), '%s\n%s\n%s'%(label,fle,label2))
+        except Exception as e: 
+            log("pbhook failed! %s" + str(e), xbmc.LOGERROR)
+            dia.update(100)
+        if dia.iscanceled(): raise Exception
+
+
+    def installEXE(self, exefile):
+        if not xbmcvfs.exists(exefile): return
+        xbmc.executebuiltin('XBMC.AlarmClock(shutdowntimer,XBMC.Quit(),2.0,true)')
+        self.killKodi.start()
+        subprocess.call(exefile, shell=True)
+        
+        
+    def killME(self):
+        subprocess.call('taskkill /f /im kodi.exe')
+        
+        
+if __name__ == '__main__': Installer()
```

Comment:
would be nice if you could move the installer somewhere else to make the plugin faster